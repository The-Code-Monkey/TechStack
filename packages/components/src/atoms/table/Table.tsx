import {
  flexRender,
  getCoreRowModel as getCoreRowModelFn,
  getSortedRowModel as getSortedRowModelFn,
  SortingState,
  TableOptions,
  useReactTable,
} from '@tanstack/react-table';
import { FC, useContext, useRef, useState } from 'react';
import { useVirtual } from 'react-virtual';
import { ThemeContext } from 'styled-components';

import { DefaultThemeWithDefaultStyles } from '../../utils';

import {
  StyledTable,
  StyledTHead,
  StyledTr,
  StyledTh,
  StyledTBody,
  StyledTd,
} from './styled';

export interface Props
  extends Omit<
    TableOptions<Record<string, unknown>>,
    'onSortingChange' | 'getCoreRowModel' | 'columns'
  > {
  className?: string;
  onRowClick?: (id: string | number) => void;
  getCoreRowModel?: TableOptions<Record<string, unknown>>['getCoreRowModel'];
  columns: Array<string>;
}

const idVariations = ['id', 'Id', 'ID'];

const Table: FC<Props> = ({
  columns,
  data,
  className,
  getCoreRowModel = getCoreRowModelFn(),
  getSortedRowModel = getSortedRowModelFn(),
}: Props) => {
  const theme = useContext<DefaultThemeWithDefaultStyles>(ThemeContext);

  const [sorting, setSorting] = useState<SortingState>([]);

  const table = useReactTable({
    data,
    columns: columns.map(column =>
      idVariations.includes(column)
        ? { header: 'ID', accessorKey: column, size: 60 }
        : { header: column, accessorKey: column }
    ),
    state: {
      sorting,
    },
    onSortingChange: setSorting,
    getCoreRowModel,
    getSortedRowModel,
    debugTable: process.env.NODE_ENV !== 'PRODUCTION',
  });

  const tableContainerRef = useRef<HTMLDivElement>(null);

  const { rows } = table.getRowModel();
  const rowVirtualizer = useVirtual({
    parentRef: tableContainerRef,
    size: rows.length,
    overscan: 10,
  });
  const { virtualItems: virtualRows, totalSize } = rowVirtualizer;

  const paddingTop = virtualRows.length > 0 ? virtualRows?.[0]?.start || 0 : 0;
  const paddingBottom =
    virtualRows.length > 0
      ? totalSize - (virtualRows?.[virtualRows.length - 1]?.end || 0)
      : 0;

  console.log(theme);

  return (
    <div ref={tableContainerRef}>
      <StyledTable defaultStyles={'table'} className={className}>
        <StyledTHead defaultStyles={'thead'}>
          {table.getHeaderGroups().map(headerGroup => (
            <StyledTr key={headerGroup.id} defaultStyles={'theadTr'}>
              {headerGroup.headers.map(header => {
                return (
                  <StyledTh
                    key={header.id}
                    colSpan={header.colSpan}
                    style={{ width: header.getSize() }}
                    defaultStyles={'th'}
                  >
                    {header.isPlaceholder ? null : (
                      <div
                        {...{
                          className: header.column.getCanSort()
                            ? 'cursor-pointer select-none'
                            : '',
                          onClick: header.column.getToggleSortingHandler(),
                        }}
                      >
                        {flexRender(
                          header.column.columnDef.header,
                          header.getContext()
                        )}
                        {{
                          asc: ' ðŸ”¼',
                          desc: ' ðŸ”½',
                        }[header.column.getIsSorted() as string] ?? null}
                      </div>
                    )}
                  </StyledTh>
                );
              })}
            </StyledTr>
          ))}
        </StyledTHead>
        <StyledTBody defaultStyles={'tbody'}>
          {paddingTop > 0 && (
            <StyledTr defaultStyles={'tbodyTr'}>
              <StyledTd
                style={{ height: `${paddingTop}px` }}
                defaultStyles={'td'}
              ></StyledTd>
            </StyledTr>
          )}
          {virtualRows.map(virtualRow => {
            const row = rows[virtualRow.index];
            return (
              <StyledTr defaultStyles={'tbodyTr'} key={row.id}>
                {row.getVisibleCells().map(cell => {
                  return (
                    <StyledTd key={cell.id} defaultStyles={'td'}>
                      {flexRender(
                        cell.column.columnDef.cell,
                        cell.getContext()
                      )}
                    </StyledTd>
                  );
                })}
              </StyledTr>
            );
          })}
          {paddingBottom > 0 && (
            <StyledTr defaultStyles={'tbodyTr'}>
              <StyledTd
                defaultStyles={'td'}
                style={{ height: `${paddingBottom}px` }}
              />
            </StyledTr>
          )}
        </StyledTBody>
      </StyledTable>
    </div>
  );
};

export default Table;
