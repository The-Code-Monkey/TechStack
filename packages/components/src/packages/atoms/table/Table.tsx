import cn from 'classnames';
import React, { FC, useCallback, useContext, useMemo } from 'react';
import { Accessor, TableOptions, useTable } from 'react-table';
import { ThemeContext } from 'styled-components';

import { ITheme } from '../../utils';

import {
  StyledTable,
  StyledTHead,
  StyledTr,
  StyledTh,
  StyledTBody,
  StyledTd,
} from './styled';

export interface Props extends TableOptions<object> {
  className?: string;
  onRowClick?: (id: string | number) => void;
}

const Table: FC<Props> = ({ columns, data, className, onRowClick }: Props) => {
  const theme = useContext<ITheme>(ThemeContext);
  const { getTableProps, getTableBodyProps, headerGroups, rows, prepareRow } =
    useTable({ columns, data });

  const tableProps = useMemo(() => getTableProps(), [getTableProps]);

  const handleRowClick = useCallback(
    (id: any) => {
      if (onRowClick) onRowClick(id);
    },
    [onRowClick]
  );

  const handleColumnClick = useCallback(
    (
      column:
        | Accessor<object>
        | (string & (string | Accessor<object> | undefined))
        | undefined
    ) => {
      console.log('Sort Column: ', column);
    },
    []
  );

  return (
    <StyledTable
      {...tableProps}
      {...theme.defaultStyles.table}
      className={cn(className, tableProps.className)}
    >
      <StyledTHead {...theme.defaultStyles.thead}>
        {headerGroups.map((headerGroup) => (
          <StyledTr
            {...headerGroup.getHeaderGroupProps()}
            {...theme.defaultStyles.tr}
            {...theme.defaultStyles.theadTr}
          >
            {headerGroup.headers.map((column, index) => (
              <StyledTh
                {...column.getHeaderProps()}
                {...theme.defaultStyles.th}
                onClick={() => handleColumnClick(columns[index].accessor)}
              >
                {column.render('Header')}
              </StyledTh>
            ))}
          </StyledTr>
        ))}
      </StyledTHead>
      <StyledTBody {...getTableBodyProps()} {...theme.defaultStyles.tbody}>
        {rows.map((row) => {
          prepareRow(row);
          return (
            <StyledTr
              {...row.getRowProps()}
              {...theme.defaultStyles.tr}
              {...theme.defaultStyles.tbodyTr}
              onClick={() => handleRowClick(row.id)}
            >
              {row.cells.map((cell) => {
                return (
                  <StyledTd
                    {...cell.getCellProps()}
                    {...theme.defaultStyles.td}
                  >
                    {cell.render('Cell')}
                  </StyledTd>
                );
              })}
            </StyledTr>
          );
        })}
      </StyledTBody>
    </StyledTable>
  );
};

export default Table;
