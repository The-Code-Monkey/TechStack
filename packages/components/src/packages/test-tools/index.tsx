import { mount, shallow, ShallowWrapper, ReactWrapper } from 'enzyme';
import React, { ReactNode } from 'react';
import { ThemeProvider } from 'styled-components';

import theme from '../theme';
import { ConfigContext } from '../utils';

const config = {
  outputDir: '/src/packages/',
  iconDir: 'feather',
};

const removeProperties = () => {
  const keys = ['theme', 'styledTheme'];
  return {
    test: (val: any) =>
      val &&
      typeof val === 'object' &&
      'props' in val &&
      Object.keys(val.props).some((prop) => keys.some((key) => key === prop)),
    print: (val: any, serialize: any) => {
      keys.forEach((key) => {
        delete val.props[key];
      });
      return serialize(val);
    },
  };
};

const formattedTheme = {
  ...theme,
  colors: {
    common: theme.colors?.common,
    modes: theme.colors?.modes,
    ...theme.colors?.modes.light,
  },
};

export const shallowWithTheme: (children: ReactNode) => ShallowWrapper = (
  children: ReactNode
) =>
  shallow(<ThemeProvider theme={formattedTheme}>{children}</ThemeProvider>)
    .dive()
    .shallow();

export const mountWithTheme: (children: ReactNode) => ReactWrapper = (
  children: ReactNode
) => {
  expect.addSnapshotSerializer(removeProperties());

  return mount(
    <ConfigContext.Provider value={config}>
      <ThemeProvider theme={formattedTheme}>
        <>{children}</>
      </ThemeProvider>
    </ConfigContext.Provider>
  );
};
