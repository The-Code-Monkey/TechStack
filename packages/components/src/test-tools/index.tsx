import isPropValid from '@emotion/is-prop-valid';
import { render, RenderResult } from '@testing-library/react';
import { ReactNode } from 'react';
import { StyleSheetManager, ThemeProvider } from 'styled-components';
import { expect } from 'bun:test';

import theme from '../theme';
import { ConfigContext } from '../utils';

const config = {
  outputDir: '/src/packages/',
  iconDir: 'feather',
};

const removeProperties = () => {
  const keys = ['theme', 'styledTheme'];
  return {
    test: (val: { props: unknown }) =>
      val &&
      typeof val === 'object' &&
      'props' in val &&
      Object.keys(val.props).some(prop => keys.some(key => key === prop)),
    serialize: (val, config, indentation, depth, refs, printer) => {
      keys.forEach(key => {
        delete val.props[key];
      });
      return printer(val, config, indentation, depth, refs);
    },
  };
};

const formattedTheme = {
  ...theme,
  colors: {
    common: theme.colors?.common,
    modes: theme.colors?.modes,
    ...theme.colors?.modes.light,
  },
};

export const mountWithTheme = (children: ReactNode): RenderResult => {
  // expect.addSnapshotSerializer(removeProperties());

  return render(
    <StyleSheetManager shouldForwardProp={isPropValid}>
      <ConfigContext.Provider value={config}>
        <ThemeProvider theme={formattedTheme}>
          <>{children}</>
        </ThemeProvider>
      </ConfigContext.Provider>
    </StyleSheetManager>
  );
};
