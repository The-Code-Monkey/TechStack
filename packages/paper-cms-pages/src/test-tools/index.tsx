import { ConfigContext } from '@techstack/components';
import { render } from '@testing-library/react';
import { ReactNode } from 'react';

import theme from '../theme';
import { ThemeProvider } from '../workarounds/styled-components';

const config = {
  outputDir: '/src/packages/',
  iconDir: 'feather',
};

const removeProperties = () => {
  const keys = ['theme', 'styledTheme'];
  return {
    test: (val: { props: unknown }) =>
      val &&
      typeof val === 'object' &&
      'props' in val &&
      Object.keys(val.props).some(prop => keys.some(key => key === prop)),
    serialize: (val, config, indentation, depth, refs, printer) => {
      keys.forEach(key => {
        delete val.props[key];
      });
      return printer(val, config, indentation, depth, refs);
    },
  };
};

const formattedTheme = {
  ...theme,
  colors: {
    common: theme.colors?.common,
    modes: theme.colors?.modes,
    ...theme.colors?.modes.light,
  },
};

export const mountWithTheme = (children: ReactNode) => {
  expect.addSnapshotSerializer(removeProperties());

  return render(
    <ConfigContext.Provider value={config}>
      <ThemeProvider theme={formattedTheme}>
        <>{children}</>
      </ThemeProvider>
    </ConfigContext.Provider>
  );
};
