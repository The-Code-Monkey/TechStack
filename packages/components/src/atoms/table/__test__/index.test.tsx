import toJson from 'enzyme-to-json';
import React from 'react';

import { mountWithTheme } from '../../../test-tools';
import Table from '../index';

describe('<Table />', () => {
  it('should render correctly', () => {
    const { asFragment } = mountWithTheme(<Table columns={[]} data={[]} />);

    expect(asFragment()).toMatchSnapshot();
  });
});
