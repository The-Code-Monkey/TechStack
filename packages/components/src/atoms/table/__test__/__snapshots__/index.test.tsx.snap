// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Table /> should render correctly 1`] = `
<DocumentFragment>
  <div>
    <table
      class="sc-jSFipO fIIapc"
    >
      <thead
        class="sc-gKAaef hIzSzz"
      >
        <tr
          class="sc-iCoHzw hElyiE"
        />
      </thead>
      <tbody
        class="sc-pNWRh iWbApb"
      />
    </table>
  </div>
</DocumentFragment>
`;
